File: crates/emu198x-commodore-web/example/parity.js
// Shared Node/browser checker. It deliberately uses only the public binding.
export function runCheck(Commodore, kind, roms, expected) {
  const check = (condition, message) => { if (!condition) throw new Error(message); };
  let machine;
  const create = () => kind === 'c64'
    ? Commodore.c64(...roms.map(bytes=>new Uint8Array(bytes)))
    : Commodore.amiga(new Uint8Array(roms[0]));
  for (let repeat=0; repeat<2; repeat++) {
    machine = create(); machine.configure_audio(48000);
    let samples=0, energy=0;
    try {
      check(machine.advance(NaN)===0,'NaN clock input must not advance');
      check(machine.advance(-10)===0,'Negative clock input must not advance');
      let rejected=false;
      try { machine.load('invalid',new Uint8Array([0])); } catch { rejected=true; }
      check(rejected,'Malformed media was accepted');
      for(let frame=1;frame<=360;frame++) {
        if(frame===301) {machine.key('a',true);machine.joystick('fire',true);machine.mouse_move(20,-10);machine.mouse_button('left',true);}
        if(frame===311) {machine.key('a',false);machine.joystick('fire',false);machine.mouse_button('left',false);}
        machine.step();
        const audio=machine.audio(); samples+=audio.length;
        for(const value of audio) {check(Number.isFinite(value),'Non-finite audio');energy+=value*value;}
        const target=expected.checkpoints.find(x=>x.frame===frame);
        if(!target)continue;
        let hash=2166136261;
        for(const byte of machine.pixels())hash=Math.imul(hash^byte,16777619)>>>0;
        check(hash.toString(16).padStart(8,'0')===target.hash,`Frame ${frame}: pixel mismatch`);
        check(machine.width()===target.width && machine.height()===target.height,`Frame ${frame}: dimensions differ`);
        check(samples===target.samples,`Frame ${frame}: sample count ${samples} != ${target.samples}`);
        // Floating-point filters need not be bit-identical across native and WASM.
        check(Math.abs(energy-target.energy)<=Math.max(1e-6,Math.abs(target.energy)*1e-5),`Frame ${frame}: audio energy differs`);
      }
    } finally {machine.free();}
  }
  let rejected=false;
  try {Commodore.amiga(new Uint8Array(1));} catch {rejected=true;}
  check(rejected,'Invalid Kickstart accepted');
  return `${kind}: native/WASM frame, sample-count and audio-energy parity at 1/100/300/310/360 frames; two fresh boots`;
}

File: crates/emu198x-commodore-web/examples/parity.rs
//! Native checkpoints for scripts/check-parity.mjs. Firmware remains local.
use emu198x_commodore_web::Commodore;
use serde_json::json;

fn hash(bytes: &[u8]) -> String {
    format!(
        "{:08x}",
        bytes.iter().fold(2_166_136_261_u32, |h, b| {
            (h ^ u32::from(*b)).wrapping_mul(16_777_619)
        })
    )
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<_> = std::env::args().skip(1).collect();
    let kind = args
        .first()
        .ok_or("pass c64 KERNAL BASIC CHARGEN [1541] or amiga KICKSTART")?;
    let mut machine = match kind.as_str() {
        "c64" if args.len() == 4 || args.len() == 5 => Commodore::c64(
            &std::fs::read(&args[1])?,
            &std::fs::read(&args[2])?,
            &std::fs::read(&args[3])?,
            &if args.len() == 5 {
                std::fs::read(&args[4])?
            } else {
                vec![]
            },
        )?,
        "amiga" if args.len() == 2 => Commodore::amiga(&std::fs::read(&args[1])?)?,
        _ => return Err("pass c64 KERNAL BASIC CHARGEN [1541] or amiga KICKSTART".into()),
    };
    machine.configure_audio(48_000)?;
    let mut checkpoints = Vec::new();
    let mut samples = 0usize;
    let mut energy = 0.0_f64;
    for frame in 1..=360 {
        if frame == 301 {
            machine.key("a", true);
            machine.joystick("fire", true)?;
            machine.mouse_move(20, -10);
            machine.mouse_button("left", true);
        }
        if frame == 311 {
            machine.key("a", false);
            machine.joystick("fire", false)?;
            machine.mouse_button("left", false);
        }
        machine.step()?;
        let audio = machine.audio();
        assert!(audio.iter().all(|v| v.is_finite()));
        samples += audio.len();
        energy += audio.iter().map(|v| f64::from(*v).powi(2)).sum::<f64>();
        if [1, 100, 300, 310, 360].contains(&frame) {
            checkpoints.push(json!({"frame":frame,"hash":hash(&machine.pixels()),
                "width":machine.width(),"height":machine.height(),"samples":samples,"energy":energy}));
        }
    }
    println!("{}", json!({"kind":kind,"checkpoints":checkpoints}));
    Ok(())
}
